torch/distributed/internal_rpc_utils.py

```diff
@@ -5,28 +5,60 @@
 import io
 import pickle
 import traceback
+import threading
+import sys
 
+import torch
+
+_thread_local_tensor_tables = threading.local()
+_thread_local_tensor_tables.recv_tables = []
+_thread_local_tensor_tables.send_tables = []
+
+
+def _tensor_receiver(tensor_index):
+    global _thread_local_tensor_tables
+    return _thread_local_tensor_tables.recv_tables[tensor_index]
+
+def _tensor_reducer(obj):
+    global _thread_local_tensor_tables
+    _thread_local_tensor_tables.send_tables.append(obj)
+    tensor_index = len(_thread_local_tensor_tables.send_tables) - 1
+    return (_tensor_receiver, (tensor_index, ))
+
+if sys.version_info >= (3, 0):
+    _dispatch_table = copyreg.dispatch_table.copy()
+    _dispatch_table[torch.Tensor] = _tensor_reducer
 
 def serialize(obj):
     f = io.BytesIO()
     p = pickle.Pickler(f)
-    p.dispatch_table = copyreg.dispatch_table.copy()
+    global _dispatch_table
+    p.dispatch_table = _dispatch_table
+    global _thread_local_tensor_tables
+    _thread_local_tensor_tables.send_tables = []
     p.dump(obj)
     return f.getvalue()
 
-
-def run_python_udf_internal(pickled_python_udf):
+# internal python function will be imported and executed in C++ land
+# it unpickles pickled python udf strings and tensors and run the python
+# udf, return serialized result and tensor tables
+def run_python_udf_internal(pickled_python_udf, tensors):
+    global _thread_local_tensor_tables
+    _thread_local_tensor_tables.recv_tables = tensors
```

Comment:
I believe nested deserialization could happen in the same thread too -- if for example __setstate__ in a class actually calls into rpc -- see comment above.